src/crypto/verification/QRCode.js

```diff
@@ -21,103 +22,59 @@ limitations under the License.
+    static get NAME() {
+        return "m.reciprocate.v1";
     }
 
     async _doVerification() {
-        const code = await new Promise((resolve, reject) => {
-            this.emit("scan", {
-                done: resolve,
-                cancel: () => reject(newUserCancelledError()),
-            });
-        });
-
-        const match = code.match(MATRIXTO_REGEXP);
-        let deviceId;
-        const keys = {};
-        if (!match) {
-            throw newQRCodeError();
-        }
-        const userId = match[1];
-        const params = match[2].split("&").map(
-            (x) => x.split("=", 2).map(decodeURIComponent),
-        );
-        let action;
-        for (const [name, value] of params) {
-            if (name === "device") {
-                deviceId = value;
-            } else if (name === "action") {
-                action = value;
-            } else {
-                const keyMatch = name.match(KEY_REGEXP);
-                if (keyMatch) {
-                    keys[keyMatch[1]] = value;
-                }
-            }
-        }
-        if (!deviceId || action !== "verify" || Object.keys(keys).length === 0) {
-            throw newQRCodeError();
+        if (!this.startEvent) {
+            // TODO: Support scanning QR codes
+            throw new Error("It is not currently possible to start verification" +
+                "with this method yet.");
         }
 
+        const targetUserId = this.startEvent.getSender();
         if (!this.userId) {
-            await new Promise((resolve, reject) => {
+            console.log("Asking to confirm user ID");
+            this.userId = await new Promise((resolve, reject) => {
                 this.emit("confirm_user_id", {
-                    userId: userId,
-                    confirm: resolve,
+                    userId: targetUserId,
+                    confirm: resolve, // takes a userId
                     cancel: () => reject(newUserMismatchError()),
                 });
             });
-        } else if (this.userId !== userId) {
+        } else if (targetUserId !== this.userId) {
             throw newUserMismatchError({
                 expected: this.userId,
-                actual: userId,
+                actual: targetUserId,
             });
         }
 
-        await this._verifyKeys(userId, keys, (keyId, device, key) => {
-            if (device.keys[keyId] !== key) {
-                throw newKeyMismatchError();
-            }
-        });
+        if (this.startEvent.getContent()['secret'] !== this.request.encodedSharedSecret) {
+            throw newKeyMismatchError();
+        }
+
+        // If we've gotten this far, verify the user's master cross signing key
+        const xsignInfo = this._baseApis.getStoredCrossSigningInfo(this.userId);
+        if (!xsignInfo) throw new Error("Missing cross signing info");
+
+        const masterKey = xsignInfo.getId("master");
+        await this._verifyKeys(this.userId, [masterKey, masterKey]);
```

Comment:
I'm not sure I really get what's going on here - presumably trying to get it into the right format so _verifyKeys can do Object.entries() but why twice? And do we not want to also verify device keys?

Reply:
we don't know the device keys at this point anymore because of how our verification setup is done.

As far as I'm aware, the format is `[keyId, key]`, and the master key's ID is the key. This should probably be `[[masterKey, masterKey]]` though

Reply:
ok, thought it would probably be something like that